frb_rust/src/ffi.rs

```diff
@@ -0,0 +1,214 @@
+//! Stubs for allo_isolate on the web.
+
+#[cfg(not(target_family = "wasm"))]
+pub use allo_isolate::{ffi::DartCObject, *};
+#[cfg(not(target_family = "wasm"))]
+pub type MessagePort = i64;
+
+#[macro_export]
+macro_rules! console_log {
+    ($lit:literal) => {
+        #[allow(unused_unsafe)]
+        unsafe { $crate::ffi::log($lit) }
+    };
+    ($( $tt:tt )*) => {
+        #[allow(unused_unsafe)]
+        unsafe { $crate::ffi::log(&format!($( $tt )*)) }
+    };
+}
+
+#[macro_export]
+macro_rules! console_error {
+    ($lit:literal) => {
+        #[allow(unused_unsafe)]
+        unsafe { $crate::ffi::error($lit) }
+    };
+    ($( $tt:tt )*) => {
+        #[allow(unused_unsafe)]
+        unsafe { $crate::ffi::error(&format!($( $tt )*)) }
+    };
+}
+
+#[cfg(target_family = "wasm")]
+pub use web::*;
+#[cfg(target_family = "wasm")]
+mod web {
+    use js_sys::{global, Array};
+    use std::iter::FromIterator;
+    pub use wasm_bindgen::closure::Closure;
+    use wasm_bindgen::prelude::*;
+    use wasm_bindgen::JsCast;
+    use web_sys::{DedicatedWorkerGlobalScope, Worker};
+    pub type MessagePort = web_sys::MessagePort;
+
+    pub type DartCObject = JsValue;
+
+    pub trait IntoDart {
+        fn into_dart(self) -> DartCObject;
+    }
+
+    pub trait IntoDartExceptPrimitive: IntoDart {}
+    impl IntoDartExceptPrimitive for JsValue {}
+
+    impl IntoDart for () {
+        #[inline]
+        fn into_dart(self) -> DartCObject {
+            JsValue::undefined()
+        }
+    }
+    macro_rules! delegate {
+        ($( $ty:ty )*) => {$(
+            impl IntoDart for $ty {
+                #[inline]
+                fn into_dart(self) -> DartCObject {
+                    DartCObject::from(self)
+                }
```

Comment:
A bit confused: IIRC IntoDart is already implemented for many types, so why do we write down this blanket implementation here

I guess it is about what you said "WASM shims for allo_isolate (merge upstream?)". Then I vote for merging to upstream :) Let a crate focus on one thing - allo_isolate for Dart isolate wrapper and frb for high-level

Reply:
Part of the reason why I hesitated merging to allo_isolate is because its whole point is to track the Dart C API, which doesn't exist in our web implementation. I might consider it if there is interest however.